tsdb/field_validator.go

```diff
@@ -8,18 +8,31 @@ import (
 	"github.com/influxdata/influxql"
 )
 
-// FieldValidator should return a PartialWriteError if the point should not be written.
-type FieldValidator interface {
-	Validate(mf *MeasurementFields, point models.Point) error
-}
-
-// defaultFieldValidator ensures that points do not use different types for fields that already exist.
-type defaultFieldValidator struct{}
+const MaxFieldValueLength = 1048576
 
-// Validate will return a PartialWriteError if the point has inconsistent fields.
-func (defaultFieldValidator) Validate(mf *MeasurementFields, point models.Point) error {
+// ValidateFields will return a PartialWriteError if:
+//  - the point has inconsistent fields, or
+//  - the point has fields that are too long
+func ValidateFields(mf *MeasurementFields, point models.Point, skipSizeValidation bool) error {
+	pointSize := point.StringSize()
 	iter := point.FieldIterator()
 	for iter.Next() {
+		if !skipSizeValidation {
+			// Check for size of field too large. Note it is much cheaper to check the whole point size
+			// than checking the StringValue size (StringValue potentially takes an allocation if it must
+			// unescape the string, and must at least parse the string)
+			if pointSize > MaxFieldValueLength && iter.Type() == models.String {
```

Comment:
A possible optimization here is to compare `pointSize` to `MaxFieldValueLength` before the loop, and set `skipSizeValidation = true` before starting the loop if `pointSize <= MaxFieldValueLength`. It's possible the compiler is already performing this loop invariant hoisting, but we can make sure it is. Even if the compiler is already doing the work, this might improve readability. It would look something like:

```
    pointSize := point.StringSize()
    if pointSize <= MaxFieldValueLength {
        skipSizeValidation = true
    }
    iter := point.FieldIterator()
    for iter.Next() {
        if !skipSizeValidation {
            if iter.Type() == models.String {
```

Reply:
I think the code as written is simpler, and I don't want to make it less simple for a single comparison - I doubt that would show up in any profiling.

It is debatable whether I should even be doing the pointSize optimization, but I think it is a clear enough win to not re-allocate a ton of strings during StringValue that it is ok to add the check.